Docs/chandra explain1

Chandra-IBM1 opened the pull request.

Commit:
Document update

Commit:
New changes

UtkarshIBM1 commented on docs/pages/task/explain_module.md:
```diff
@@ -59,9 +54,8 @@ Each exported procedure should be documented using the following structure:
     -  Required preconditions
     -  Unsupported input types
 
-  -  `Outcomes`
```
Since we have removed Output sections, I feel that outcomes will be a valuable section to keep as this will provide the user an overall view of what result will be produce after code execution. @bmorris1 @edmundreinhardt . Could you please suggest on this point ?

edmundreinhardt replied:
Side Effects and Return Values and Output Parameters should be sufficent to document that.

anandsha20 commented on docs/pages/task/explain_module.md:
```diff
@@ -59,9 +54,8 @@ Each exported procedure should be documented using the following structure:
 
-  -  `Outcomes`
+  -  `Usage Example`
     Define the expected outcomes of the procedure’s execution.
```
`Usage Example`: This should provide an example on how we can utilize or call the procedure instead of defining the expected outcomes. fyi @bmorris1, @edmundreinhardt, @Chandra-IBM1

bmorris1 replied:
@anandsha20 I agree that it should show how to call the procedure.

It should show an example of all aspects of calling the procedure

- The /copy statement with the prototype 
- any setup needed for parameters
- the call to the procedure
- if it returns a value or modifies a parameter, it should either show or describe how the caller would use those values.

But for cases where the caller would not do any further coding related to that call, I don't think anything is needed in the Usage section. For example, if the procedure did something related to files, that's not really part of the Usage, it's a Side Effect.

anandsha20 commented on docs/pages/task/explain_program.md:
```diff
@@ -12,15 +12,15 @@ a high-level summary of the program's purpose and behavior, detailing the parame
   - Example: The load all subfile can load only up to 9999 records.
 
-6. `Outcomes`: List possible major outcomes or scenarios resulting from the code execution. This section can include an optional `usage example` to illustrate the expected outcome.
+6. `Usage Example`: Define the expected outcomes of the program execution.
```
Usage Example: This should provide an example on how we can utilize the program instead of defining the expected outcomes. Please check for other scopes as well. fyi @bmorris1, @edmundreinhardt, @Chandra-IBM1

bmorris1 replied:
I agree. It should be similar to the Usage section for a procedure in a module, although depending on the nature of the program it might be better to show a ```clle call which could also represent a call from the command line.

But if there is a /copy file with an RPG prototype for the program, the Usage should be similar to the Usage for a procedure

- The /copy statement
- Any setup for parameters
- The call
- Code or discussion saying what to do with any output variables

edmundreinhardt replied:
I agree that this should be an example of calling the program.  The outcomes should have been discussed in side effects .
In fact for programs, I think we should rename `Side Effects` to `Outcomes`
What do you think @bmorris1 @anandsha20 @Chandra-IBM1

bmorris1 replied:
I'm not sure. 

Any procedure or program could have both outcomes and side effects, although I'm not always sure which is which.

How about having a more general name for both outcomes and side effects such as "Outside effects", although I don't really like that name. For procedures and subroutines, it could have a subsection about "Global variables".

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -37,20 +37,15 @@ Each exported procedure should be documented using the following structure:
     -  Global variables
     -  Shared resources
-    -  Logs or external states
+    -  Dataaera updates 
```
It should not discuss Global variables unless they are defined with the IMPORT or EXPORT keyword.

edmundreinhardt replied:
- Data area or DATARA

bmorris1 replied:
Data area or DTAARA

bmorris1 replied:
I think "Data area" is better

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -75,26 +69,11 @@ The how_output part explains how the module works internally, covering the full
 
-  - `Initialization`: Describe the initial setup required for the procedure. Include all local variables and parameters.
+  - `Definitions`: This section covers all declarations and setup required before the main logic of the procedure begins. It includes the definition of local variables, data structures, and any constants or initial values needed to support the procedure’s execution
   - `Main Logic`: Explain the core logic of the procedure, detailing the steps taken to achieve the procedure's purpose.
```
It should list any subroutines within the procedure with a brief description of each subroutine.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -75,26 +69,11 @@ The how_output part explains how the module works internally, covering the full
+3. `Global Components`: The Global Variables section includes all variables, prototypes for external program calls, data structures, arrays, constants, and keywords used in the variable declarations (example`LIKE`, `INZ`, `N` data type)
 
-  Provide a step-by-step description of the logic for each procedure.
+4. `Procedures` : Provide a step-by-step description of the logic for each procedure.
```
For each procedure, state whether it is exported or internal to the module.

bmorris1 commented on docs/pages/task/explain_source_references.md:
```diff
@@ -28,32 +28,21 @@ The `api_output` part provides a high-level summary of the procedure or subrouti
 
 3. `Return Values` : List the return value accepted by the procedure, specifying whether it is an output
 
-4. `Dependencies`: This section outlines all components that the procedure or subroutine depends on in order to run successfully. Dependencies include physical and logical files used for data access, external procedures for business logic, and display files for user interface interactions.
```
Git would not let me put this comment on the "Return values" line.

It should say 

3. `Return value`: List the value returned by the procedure. Specify the data type and the meaning of the return value.

bmorris1 commented on docs/pages/task/explain_source_references.md:
```diff
@@ -28,32 +28,21 @@ The `api_output` part provides a high-level summary of the procedure or subrouti
   - Example: The caller cannot update records it is only input. This means the procedure or subroutine is designed to read data but not modify it.
       
-7. `Outcomes`: List possible major outcomes or scenarios resulting from the code execution. This section can include an optional usage example to illustrate the expected outcome.
+7. `Usage Example`: List possible major outcomes or scenarios resulting from the code execution. This section can include an optional usage example to illustrate the expected outcome.
```
This should show a call to the procedure, with any setup required for parameters, and either an example or a description of what the caller would do with the return value or any output parameters.

bmorris1 commented on docs/pages/task/explain_source_references.md:
```diff
@@ -28,32 +28,21 @@ The `api_output` part provides a high-level summary of the procedure or subrouti
+  
 2. `Field Mapping`: If the procedure or subroutine relates to any display file or printer file, create a clear table or list to show how physical/logical file fields are mapped to display fields or printer file. This includes field names and display fields.
 
 3. `Indicators`: If the procedure or subroutine uses indicators, explain their purpose and how they are set or cleared during execution in table format.
```
Clarify that this is only the *IN indicators or any indicators in an INDDS data structure.

It should not be about any ordinary variables that happen to have the indicator data type.

bmorris1 commented on docs/pages/task/sample_module.md:
```diff
@@ -94,27 +94,16 @@ The `GetActInf` module provides reusable subprocedures to retrieve account-relat
 #### Return Value
 | -Return Value- | Output    | `char(10)` | The account number associated with the given ID |
```
For the Return Value, The first two columns are not needed.

| Data Type  | Description                                     |
| ---------- | ----------------------------------------------- |
| `char(10)` |  The account number associated with the given ID       |

If it is a data structure, list the subfields.

bmorris1 commented on docs/pages/task/sample_module.md:
```diff
@@ -141,28 +130,17 @@ account_number = GetAccNo(userid);
-    - The returned value will be blank.
-    - Calling code must ensure the returned value is not reused inappropriately in future calls.
-
 #### Usage Example
```
For small snippets, I don't think it would be good to put **free at the top. But without **free, free-form code has to start in column 8. So, I think any little snippet like this should have 7 blanks before the code. 

This version has the code starting in column 1. The formatter thinks the 'd' in `userid` is the `D` of a fixed-form D spec and it doesn't "see" the "dcl-s" part of the declarations so it doesn't colour them red.

```rpgle
dcl-s userid char(10);
dcl-s account_number char(10);
/copy getactinfcpy;

userid = 'TEST01';
account_number = GetAccNo(userid);
```

This version has 7 blanks preceding each line of code.

```rpgle
       dcl-s userid char(10);
       dcl-s account_number char(10);

       /copy getactinfcpy;

       userid = 'TEST01';
       account_number = GetAccNo(userid);
```

bmorris1 commented on docs/pages/task/sample_module.md:
```diff
@@ -141,28 +130,17 @@ account_number = GetAccNo(userid);
 | `P_UserId`     | Input     | `char(10)` | Customer ID to search for in the `AccPf` file              |
 | -Return Value- | Output    | `char(50)` | The account description associated with the input `CustId` |
```
Fix this information about the return value to be in the same format as for `GetAccNo` above.

bmorris1 commented on docs/pages/task/sample_module.md:
```diff
@@ -174,13 +152,11 @@ dcl-s account_desc char(50);
 - If no match is found:
   `account_desc` will be blank. The calling logic should handle such cases to avoid misinterpretation.
 
-
```
Git would not let me add a comment on the line about the purpose. 

It currently says 

> The RPGLE program provides reusable subprocedures to retrieve account-related information from the `AccPf` physical file based on a given customer ID (`CustId`). Specifically, it allows external programs to obtain:

It should say 

> The RPGLE module provides reusable subprocedures to retrieve account-related information from the `AccPf` physical file based on a given customer ID (`CustId`). Specifically, it contains exported procedures to obtain:

bmorris1 commented on docs/pages/task/sample_program.md:
```diff
@@ -41,33 +41,24 @@ Therefore the `context` directory would look like:
 
 The RPGLE program is designed to retrieve the account number (`AccNo`) for a given user ID (`P_UserId`) from the `AccPf` file. It reads through the records in the `AccPf` file and searches for a match with the provided user ID. Once a match is found, it sets the corresponding account number in the output parameter `P_AccNo`.
 
-### Parameters
```
Git would not let me add a comment at the beginning of this file. The code sample for the program must have the **free right at the beginning of the line. 

This is what the code looks like with the **free starting in column 5. It thinks the code is column-limited, so it interprets column 6 as being the spec type.

```rpgle
    **free
    ctl-opt dftactgrp(*no) actgrp(*new);

    dcl-f AccPf if e k disk;
```

This is what it looks like with the **free in column 1:

```rpgle
**free
    ctl-opt dftactgrp(*no) actgrp(*new);

    dcl-f AccPf if e k disk;
```

bmorris1 replied:
This comment was not addressed yet.

bmorris1 commented on docs/pages/task/sample_program.md:
```diff
@@ -41,33 +41,24 @@ Therefore the `context` directory would look like:
-  - If the same program runs again without resetting `P_AccNo`, it retains the previous value, leading to incorrect or unintended output.
-
-#### Usage Example
+### 6. Usage Example
```
Git would not let me add a comment in the File Specifications section.

It has this bit of sample code for a fixed-form statement. This would not be helpful for someone who does not know RPG.

 `FAccPf IF E K Disk`
 
Please correct all such samples of fixed-form statements in all these documents to show the actual statement, surrounded by the two lines ```rpgle and ``` and with 5 leading blanks.
  
 ```rpgle
      FAccPf     IF   E      K      Disk

bmorris1 replied:
I think it would be good to have a page about how to format code snippets, and link to the page from each document.

- For small free-form snippets, don't put **free, but instead have 7 leading spaces
- For large free-form snippets, make sure the **free is in column 1.
- For fixed-form snippets, always have the correct fixed-format of the lines

bmorris1 reviewed (changes requested):
Some of my comments apply to all the documents, such as my comments about formatting code snippets.

Also, some of my comments are not positioned against the line they apply to, because Git would not let me make a comment for some lines in large unchanged sections. I hope I made it clear what I was actually referring to ... :-)

Commit:
Modified as per latest call

Commit:
Modified as per latest call

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  Example:
+  | Parameter Name | Data Type | Description |
+  |----------------|------|--------|
+  | `pOrderId`     | `char(10)` | The unique identifier for the order being processed |
```
Add a column after the Data Type column for whether it is input or output. 

- If CONST or VALUE is specified, it is input-only
- If CONST or VALUE is not specified, it is input and output

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  - APIs or external libraries
+  - Data areas
+  - Data queues
+  - Any other external modules or systems
```
Change the last one to just "etc"

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  List all external components the program relies on to function correctly. This includes:
+  - Called programs
+  - Service programs (SRVPGMs)
+  - APIs or external libraries
```
This should be two separate points

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  | Component Name  | Type     | Description |
+  |-----------------|----------|-------------|
+  | `CALC_SERVICE`   | Service  | Performs complex calculations for order processing |
+  | `DATA_AREA_1`    | Data Area| Stores configuration settings for the program |
```
Make this example library qualified since the library should be listed if it is known

 | `LIB1/DATA_AREA_1`    | Data Area| Stores configuration settings for the program |

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  This section should illustrate all relevant aspects of calling the procedure, including:
+
+  1. The `/COPY` Statement
+      - Show how the caller includes the prototype using a `/COPY` directive (if applicable).
```
It is always applicable for an rpgle example for scope=module, since the caller is expected to be in a different module.

bmorris1 replied:
Add this:

When there is no /COPY file with the prototypes, it is a problem in the application, especially if the other programs in the application define their own prototypes so they can call the procedures.

In that case, the issue with the missing /COPY with the prototypes should be mentioned in the "Problems" section.

bmorris1 replied:
@edmundreinhardt @Chandra-IBM1 @jiatianzhong How should usage samples be handled when the module doesn't have a copy file with the prototypes for its procedures? 

- Should the sample just make up the name of a /COPY file?
- Should the sample actually code the prototype (which would add to the problem of the prototype being coded in multiple modules)

bmorris1 replied:
I started a Slack thread about this in the #wca4i-tcs-datagen channel: https://my.slack.com/archives/C08FGHCDT26/p1750425339221139

edmundreinhardt replied:
I am unable to follow that slack link but my opinion is that if no copybook is present/known, the prototype should be in the usage example.  The developer can then copy it into a copybook
@bmorris1 @Chandra-IBM1 @jiatianzhong

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+      - Show how the caller includes the prototype using a `/COPY` directive (if applicable).
+
+  2. Parameter Setup
+    - Define and initialize all input and output parameters required for the call.
```
input and output parameters and any return value

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+
+  3. The Procedure Call
+    - Demonstrate the call using:
+      - `CALLP` if using a procedure with a prototype.
```
The CALLP opcode is not necessary. It is very rare that CALLP is used for a prototyped call in free form.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+    - Demonstrate the call using:
+      - `CALLP` if using a procedure with a prototype.
+    - If the procedure returns a value, show how the caller captures and uses that return value.
+    - If parameters are modified, show how the caller accesses the updated values.
```
Change to

- Show how the caller uses the return value or output parameters, if any

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+      DCL-S orderTotal   PACKED(9:2);
+      DCL-S statusCode   CHAR(2);
+
+      CALLP CalculateOrderTotal(customerId : orderTotal : statusCode);
```
Remove the CALLP. Just do the procedure call.

```rpgle
       CalculateOrderTotal(customerId : orderTotal : statusCode);
```

bmorris1 commented on docs/pages/task/explain_module.md:
````diff
@@ -3,98 +3,216 @@
+
+  example:
+  ```rpgle
+      /COPY QRPGLESRC,MYPROTOTYPE
````
All the free-form snippets must be indented by at least 7 spaces before the code. 

This should be an IMPORTANT guideline at the beginning of each of the "explain_*" documents.

(And also fix up all the snippets in these documents. :-)

If you look at the rendered markdown you can see that it doesn't understand the code properly. Here is what yours currently looks like with fewer than 7 spaces. It thinks column 5 is the fixed-form specification type.

  ```rpgle
      /COPY QRPGLESRC,MYPROTOTYPE
      DCL-S customerId   CHAR(10)   INZ('CUST001');
      DCL-S orderTotal   PACKED(9:2);
      DCL-S statusCode   CHAR(2);
      CALLP CalculateOrderTotal(customerId : orderTotal : statusCode);      

      // After call, orderTotal and statusCode will have updated values
  ```      
    Here it is with 7 spaces:
  
   ```rpgle
          /COPY QRPGLESRC,MYPROTOTYPE
          DCL-S customerId   CHAR(10)   INZ('CUST001');
          DCL-S orderTotal   PACKED(9:2);
          DCL-S statusCode   CHAR(2);
          CALLP CalculateOrderTotal(customerId : orderTotal : statusCode);      

          // After call, orderTotal and statusCode will have updated values
  ```

bmorris1 replied:
Add an IMPORTANT guideline at the top of every explain_x file:

#### IMPORTANT
- All the free-form snippets must be indented by at least 7 spaces before the code. 
- All fixed-form snippets must have exactly 5 spaces before the specification type.
- If the ```rpgle to start the snippet is indented, the 5 or 7 spaces must start at the same indentation as the ```rpgle line.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+   | File Name       | Type     | Used | Description |
+   |-----------------|----------|------|-------------|
+   | `CUSTOMER`      | Data     | Input  | Reads customer records for processing |
+   | `ORDER_DISPLAY` | Display  | Input/Output | Displays order details to the user |
```
Add another example with "Input/Update" in the "Used" column. For RPG programmers, "output" only means the equivalent of the SQL "INSERT" statement.

   | `ORDERS` | Data | Input/Update| Updates an order|

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
-The how_output part explains how the module works internally, covering the full execution flow and logic used at each step. It includes details on control specifications, file declarations, global variables, constants, indicators, field mapping, subroutines, error handling, and the logic of each procedure.
+  Provide the subfields in a table format with columns for subfield name, type, and description.
+  Example: Data Structures
+  Subfield Name | Type    | Description |
```
For the example, put the name of the data structure before the list of subfields.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
 
-## how_output
+  ##### Data Structures
+  This section documents all data structures used in the program, including their subfields and purposes.
```
... and anything else that is interesting about the data structure such as whether it is qualified or whether it is an array.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  Subfield Name | Type    | Description |
+  |----------------|--------|-------------|
+  | `OrderDetail`   | 10 Length Character   | Details of the order |
+  | `OrderAmount`   | 7 numeric 2 decimals | Amount of the order |
```
I think there should be a standard way of documenting the data types, in a separate file that is linked from anywhere that it mentions types.

And then all the examples in these documents should follow that standard.

I think an extra meeting is needed, with all the participants.

But for now, I think the first thing mentioned should be the type, then any other information

- Character, length 10
- Packed numeric, length 7, with 2 decimals
- Date, *ISO format

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,98 +3,216 @@
+  Example: Arrays
+  | Array Name | Dimensions | Description |
+  |-------------|------------|-------------|
+  | `OrderArray` | 1000 elements | Array to hold order details |
```
Data type is also needed for arrays.

These tables would be the same as the variables and data structure tables above with an extra column for the dimension.

But I think it would be better to document arrays with the other variables and data structures. If it's an array, that could just be mentioned as part of the description.

Also, since subfields can be arrays, it would be very complex to list array subfields somehow separate from the data structure.

So for example, in the Variables table above, you could have an example like this:

 | `MeetingDates`   | Date, *ISO format  | Array with dimension 10, the available dates for meetings |

bmorris1 reviewed (commented):
I have only reviewed explain_module.md so far since the meeting on May 26/May 27.

bmorris1 commented on docs/pages/task/explain_procedure.md:
````diff
@@ -0,0 +1,187 @@
+This should show how to call the procedure, including any necessary setup for parameters, along with an example or description of how the caller would handle the return value or any output parameters.
+
+```rpgle
+     /COPY QRPGLESRC,MYPROTOTYPE
````
It is not necessary to have the /COPY for a scope=proc usage example. For this scope, the assumption is that the call will be in the same module, so the prototype for the call isn't even necessary.

bmorris1 commented on docs/pages/task/explain_procedure.md:
```diff
@@ -0,0 +1,187 @@
+  Example:
+  | Parameter Name | Data Type | Description |
+  |----------------|------|--------|
+  | `pOrderId`     | `char(10)` | The unique identifier for the order being processed |
```
I like this way of showing the data type much better than "length 10 character" and also better than "character, length 10".

Chandra-IBM1 replied:
If any variable is declared using the LIKE keyword to refer to a field from a file, and that file's DDS (Data Description Specifications) is not present in the repository. That what we should mention as data type for that?

Chandra-IBM1 replied:
Can we follow this way for other type of variable?

| Parameter Name | Data Type             | Usage        | Description                                                  |
|----------------|------------------------|--------------|--------------------------------------------------------------|
| `customerId`   | `char(10)`             | Input/Output | The unique identifier for the customer                       |
| `orderTotal`   | `packed(9:2)`          | Input/Output | The total amount of the order calculated                     |
| `currencyCode` | `char(3)` `CONST`      | Input only   | The currency in which the order is placed                    |
| `taxRate`      | `packed(5:2)` `VALUE`  | Input only   | The applicable tax rate for the order                        |
| `orderDate`    | `date(*ISO)`           | Input/Output | The date the order was placed                                |

or should we follow this way 
  | Name           | Type                                 | Usage        | Description                                                  |
  |----------------|--------------------------------------|--------------|--------------------------------------------------------------|
  | `customerId`   | Character, length 10                 | Input/Output | The unique identifier for the customer                       |
  | `orderTotal`   | Packed numeric, length 9, with 2 decimals | Input/Output | The total amount of the order calculated                     |
  | `currencyCode` | Character, length 3 (`CONST`)        | Input only   | The currency in which the order is placed                    |
  | `taxRate`      | Packed numeric, length 5, with 2 decimals (`VALUE`) | Input only   | The applicable tax rate for the order                        |
  | `orderDate`    | Date, *ISO format                    | Input/Output | The date the order was placed                                |

bmorris1 replied:
I like the first way better.

If it is defined with LIKE, put both pieces of information. 

LIKE(`somefield`), date(*ISO)

edmundreinhardt replied:
OK how about
LIKE(somefield) = DATE(*ISO)

But Chandra is asking about whether we give a natural language description
I suspect that this will become very repetitive if we always do it.  The end user only needs to learn once what PACKED means.

bmorris1 replied:
> If any variable is declared using the LIKE keyword to refer to a field from a file, and that file's DDS (Data Description Specifications) is not present in the repository. That what we should mention as data type for that?

If the meaning of the LIKE keyword is not known, then just put "LIKE(name)" for the type. And then if there is some indication of the type, such as an assignment, put something in the description such as "Likely a numeric type"

bmorris1 replied:
I agree that a natural language description is not necessary for data types.

bmorris1 commented on docs/pages/task/explain_procedure.md:
```diff
@@ -0,0 +1,187 @@
+
+  - At end of this you can explain any specific keywords used in the variable declarations, such as `LIKE`, `INZ`, or `N` data type.
+
+  ##### Arrays 
```
Many of the comments I made for explain_module also apply to this document.

bmorris1 commented on docs/pages/task/explain_program.md:
```diff
@@ -3,55 +3,175 @@
 
-6. `Outcomes`: List possible major outcomes or scenarios resulting from the code execution. This section can include an optional `usage example` to illustrate the expected outcome.
+  2. The Program Call
+    - Use `CALL` to invoke a program object.
```
Change this:

2. The program call.

- If there is an RPG prototype for the program, use an rpgle snippet with the /COPY for the prototype, and any definitions needed for parameters. Do not use the CALL or CALLP opcode.
```rpgle
       /copy qrpglesrc,prototype
              
       mypgm (parameters);
```
- Otherwise, use a clle snippet to represent a call from the command line.
```clle
call mypgm parm(parameters)
```

bmorris1 commented on docs/pages/task/explain_subroutine.md:
```diff
@@ -0,0 +1,140 @@
+    | `pStatus`      | Output    | `char(10)` | The status code indicating success or failure of the procedure |
+
+3. #### File Inputs/Outputs 
+Describe how the program reads from and writes to:
```
change "program" to "subroutine" in several places in this file.

bmorris1 commented on docs/pages/task/explain_subroutine.md:
```diff
@@ -0,0 +1,140 @@
+  | `ORDER_DISPLAY` | Display  | Input/Output | Displays order details to the user |
+  | `PRINTER_FILE`  | Printer  | Output | Sends reports to the printer |
+
+4. #### Dependencies 
```
Is this section really needed for a subroutine?

Chandra-IBM1 replied:
Yes, I believe it's important to document any external programs, APIs, data areas, or data queues that the program interacts with

bmorris1 commented on docs/pages/task/explain_subroutine.md:
```diff
@@ -0,0 +1,140 @@
+Provide a short insight into the purpose of the code for the subroutine. This should focus on the business logic or functional role of the code.
+
+2. #### Global Components 
+This section lists the global elements used by the subroutine, such as indicators.
```
Don't mention indicators here. There is already too much focus on indicators.

Instead, say this:

This section lists the global elements used by the subroutine, such as variables and files.

bmorris1 commented on docs/pages/task/explain_subroutine.md:
```diff
@@ -0,0 +1,140 @@
+
+  `Include:`
+  -  `*INxx` Indicators: Standard RPG indicators (e.g., `*IN03`) or fixed-format numeric indicators controlling logic or UI.
+  -  Indicators in `INDDS` Structures: Subfields defined inside indicator data structures linked to files using `INDARA`. These control function keys, display flags, etc.
```
This is a general comment for all the explain files. Also show an example table for an indds. I think the good layout for an indds table has already been established but I don't know where it is. But I think the following would be a good way to show an INDDS.

- Indicator data structure `myIndds` for display file `MYDSPF`

  |Subfield name  | Indicator number     | Description |
  |-----------------|----------|-------------|
  | `exit`   | 03 | Indicates that the user pressed F3  |
  | `sfl_clear`    | 55| Used to clear the subfile |
  | `error`      | 99      | Used to show error messages |

bmorris1 reviewed (changes requested):
Many of my comments apply to all the explain files.

I have not looked at the sample files yet. I will review those after changes are made according to the explain documents.

Commit:
Changes are done as per review

Commit:
Changes are done as per review comments

Commit:
Modified subroutine

bmorris1 commented on docs/pages/task/explain_subroutine.md:
````diff
@@ -0,0 +1,160 @@
+Provide a code snippet showing how the subroutine is called within the program. 
+Example:
+```rpgle
+     D InputVar        S             10A   INZ('Test')  
````
Don't define any variables for calling a subroutine.

Before the EXSR, set the values for any input variables.

After the EXSR, show how to use the output variables.

```rpgle
       InputVar = 'abc';
       EXSR MySubroutine;       
       If OutputVar > 5;
        ...
```

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,101 +3,251 @@
+
+  2. Parameter Setup
+    - Define and initialize all input, output parameters and any return values required for the call.
+    - Include declarations for any necessary constants, variables, or data structures.
```
Define data structures using LIKEDS with the name of the data structure defined in the /COPY file.

bmorris1 commented on docs/pages/task/explain_module.md:
````diff
@@ -3,101 +3,251 @@
+
+  example:
+  ```rpgle
+       /COPY QRPGLESRC,MYPROTOTYPE
````
The 7 spaces before the code must start in the same column as the ```rpgle

With only 5 extra spaces

  ```rpgle
       /COPY QRPGLESRC,MYPROTOTYPE

       DCL-S customerId   CHAR(10)   INZ('CUST001');
       DCL-S orderTotal   PACKED(9:2);
       DCL-S statusCode   CHAR(2);
  ```

With 7 extra spaces  
  example:
  ```rpgle
         /COPY QRPGLESRC,MYPROTOTYPE

         DCL-S customerId   CHAR(10)   INZ('CUST001');
         DCL-S orderTotal   PACKED(9:2);
         DCL-S statusCode   CHAR(2);
  ```

bmorris1 commented on docs/pages/task/explain_program.md:
````diff
@@ -3,55 +3,193 @@
+              
+          mypgm (parameters);
+    ```
+     - Otherwise, use a clle snippet to represent a call from the command line.
````
This is rendered as part of the example. I think it is because it is indented too much.

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,101 +3,251 @@
+  - At end of this you can explain any specific keywords used in the data structure declarations, such as `LIKEDS`, `QUALIFIED`, or `INZ`.
+
+  ##### Arrays 
+  This section documents all arrays used in the module, including their dimensions and purposes.
```
I don't think it is good to have arrays documented separately from other variables. My earlier suggestion was to avoid having an extra section for arrays, and just put dimension information as part of the description. 

But if arrays will be separate, then data structure arrays should be included in the "arrays" section. Or there should be a separate "Data structure arrays" section.

bmorris1 reviewed (commented):
I added a few more comments to the explain documents.

Commit:
Changes are done as per new format

bmorris1 commented on docs/pages/task/explain_global_definitions.md:
```diff
@@ -0,0 +1,118 @@
+##### Variables 
+This section documents all variables used in the program, module, or procedure. It should be described in tabular format with columns for Name, Data Type, Description, and Attributes.
+
+Example: Variables
```
I think the "Attributes" column should be before the "Description" column. The attributes are part of the data type.

bmorris1 replied:
Maybe it is ok to have the Attribute column last, since it is an optional column.

bmorris1 commented on docs/pages/task/explain_global_definitions.md:
```diff
@@ -0,0 +1,118 @@
+Example: `OrderHeader`  
+| Name           | Data Type               | Description                                  | Attributes |
+|----------------|------------------------ |----------------------------------------------|------------|
+| `OrderID`      | char(10)                | Unique identifier for the order              |            |
```
The attributes column is needed for subfields too

bmorris1 commented on docs/pages/task/explain.md:
```diff
@@ -35,3 +35,8 @@ In either case the JSONL will contain data, where each line is of the format:
 - Use `ILE RPG` instead of `RPGLE`.
 - Do not expand `RPG` as `Report Program Generator`. Always refer to it as `RPG`.
+- Data types should follow the format:  
+  - `Character, length 10`  
```
Data types should be expressed as free-form RPG syntax:
- `CHAR(10)`
- `PACKED(7:2)`
- `DATE(*ISO)`

bmorris1 commented on docs/pages/task/explain_global_definitions.md:
````diff
@@ -0,0 +1,118 @@
+The following qualified data structure is an array with 10 elements, used to store order header information. Each element contains details about an order, including items, requirements, and address information.
+
+```rpgle
+dcl-ds order qualified;
````
It needs the DIM(10) keyword

bmorris1 commented on docs/pages/task/sample_source_references.md:
```diff
@@ -76,31 +76,28 @@ The subprocedure `GetAccNo` is responsible for retrieving the account number (`A
 
-- Physical File – `AccPf`
-  The subprocedure `GetAccNo` depends on the physical file `AccPf`, which is defined in the main program's file declaration. It is used in input mode with keyed access.
+| Return Name | Type   | Length | Direction | Description                                                                                                  |
```
All data types should be handled the same way (where the Type column also has the length, CHAR(10) etc.

bmorris1 commented on docs/pages/task/explain_procedure.md:
```diff
@@ -0,0 +1,172 @@
+| 2 | `OrderCount`  | Output       | int(5)         | Total number of orders processed                    | INZ(0) *OMIT              |
+| 3 | `info`        | Input        | Data structure | Data structure containing order details             | DIM(10) CONST             |
+|   | `info.number` |              | int(10)        | Number of orders processed in the current session   |                           |
+|   | `info.name`   |              | char(20)       | Name of the customer associated with the order      | VARYING                   |
```
If it has VARYING, the data type is VARCHAR

bmorris1 commented on docs/pages/task/explain_subroutine.md:
```diff
@@ -0,0 +1,123 @@
+1. #### Purpose
+Provide a short insight into the purpose of the code for the subroutine. This should focus on the business logic or functional role of the code.
+
+2. #### Variable I/O
```
The term "I/O" should only be used for file I/O. 

This should be called "Global variables". It is ok call the table headings "Input" and "Output"

It would even be confusing to call it "Input and output variables" since that could mean it was only about variables used in I/O file operations.

bmorris1 committed:
Update explain_global_definitions.md

Just some spelling errors for now

bmorris1 committed:
Create explain_definitions.md

bmorris1 committed:
Update explain_definitions.md

Document how to describe the type and possible attributes of a variable, subfield, or parameter.

bmorris1 committed:
Update explain_definitions.md

Update to use "Data Type" instead of just "Type" for the column heading

bmorris1 committed:
Update explain_definitions.md

bmorris1 committed:
Update explain_global_definitions.md

Removed extra sections for different types of data structures. Will later move the examples to a separate file

bmorris1 committed:
Create explain_data_structure.md

Examples of data structure explanations

bmorris1 committed:
Update explain_global_definitions.md

Most of the information was moved into other files.

bmorris1 committed:
Create explain_parameters.md

bmorris1 committed:
Update explain_global_definitions.md

bmorris1 committed:
Update explain_parameters.md

bmorris1 committed:
Update explain.md

Minor change

bmorris1 committed:
Update explain_module.md

Minor changes

bmorris1 committed:
Update explain_procedure.md

The main change is to remove the Side Effects section since that should be covered by the global variables, file I/O, and dependencies.

bmorris1 committed:
Update explain_lines.md

Minor change

bmorris1 committed:
Update explain_program.md

Minor changes

bmorris1 committed:
Update explain_subroutine.md

Minor changes

bmorris1 committed:
Update explain_copybook.md

Minor change

bmorris1 committed:
Update sample_copybook.md

bmorris1 committed:
Update sample_lines.md

Fix formatting

bmorris1 committed:
Update explain.md

bmorris1 committed:
Update sample_module.md

Some formatting changes.

bmorris1 committed:
Update sample_program.md

Code-formatting corrections, some content corrections. 

More changes are needed for it to conform to the required output format.

bmorris1 committed:
Update sample_source_references.md

Code formatting changes, some general output-format changes, more needed.

bmorris1 committed:
Update sample_module.md

Add the parameter-number column to the parm tables

bmorris1 commented on docs/pages/task/explain_program.md:
````diff
@@ -3,55 +3,115 @@
+  - Otherwise, use a clle snippet to represent a call from the command line.
+```clle 
+        call mypgm parm(parameters)
+```
````
Question (@edmundreinhardt @Chandra-IBM1 @jiatianzhong )

If it is a fixed-form program with no prototype for the call, and the parameters in the *ENTRY PLIST form, would it be better to show a fixed form RPGLE CALL rather than showing a clle-style call from the command line?

```rpgle
     C                   call      'MYPGM'
     C                   parm                    parameter
```

Chandra-IBM1 replied:
Yes, use the fixed-form RPGLE CALL with PARM to match the style and parameter-passing method of the program. It is clearer and more accurate for understanding how the program expects to be called.

bmorris1 replied:
Ok, thanks. I changed it to have 3 possibilities: 
- free-form call if there is a prototype
- else fixed-form CALL+PARM if there is a *ENTRY PLIST
- otherwise clle call

bmorris1 committed:
Update explain_definitions.md

Fixed the algorithm for determining the number of bytes in packed fields with an even number of digits

bmorris1 committed:
Update explain_program.md 

Suggest to use a fixed-form CALL if the parameters are in *ENTRY PLIST format

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,105 +3,95 @@
-    -  External APIs or modules
+  2. Parameter Setup
+    - Define and initialize all input, output parameters and any return values required for the call.
+    - Include declarations for any necessary constants, variables, or data structures.
```
There should be sections with "Possible Problems with this Code" and "Possible Improvements to this Code"

bmorris1 commented on docs/pages/task/explain_parameters.md:
```diff
@@ -0,0 +1,41 @@
+The parameters are described in a table with columns for the parameter number, `Data Type`, `Usage`, `Description`, and possibly `Attributes`.
+
+The column for the parameter number does not have a heading.
+
```
Add this:

> If it is an unnamed parameter in a prototype, put `not named` in the name column. Also, add a point in the "Possible Improvements to this Code" that it's a good idea to give names to the parameters in a prototype

bmorris1 commented on docs/pages/task/explain_definitions.md:
```diff
@@ -0,0 +1,51 @@
+| * | PROCPTR | pointer(*proc) |
+| O | CLASS(*java ...) | object(*java ... ) |
+| D | DATFMT(fmt) | date(fmt) |
+| T | TIMFMT(fmt) | time(fmt) |
```
Add some information about data structure subfields that don't have any data type. See this Slack thread in #wca4i-tcs-datagen:  https://my.slack.com/archives/C08FGHCDT26/p1750425339221139

bmorris1 replied:
Sorry. The correct link for the Slack thread about subfields with no data type is https://my.slack.com/archives/C08FGHCDT26/p1750441642809399

bmorris1 commented on docs/pages/task/explain_module.md:
```diff
@@ -3,105 +3,95 @@
 
- 2. `Exported Procedures`
-
+2. #### Exported Procedures
```
I think the general part of api_output before the procedures should have the global dependencies (files and other objects)

bmorris1 reviewed (commented):
I made several new comments but I haven't finished my review yet.

bmorris1 commented on docs/pages/task/explain_global_definitions.md:
```diff
@@ -0,0 +1,30 @@
+
+##### Variables 
+This section documents all variables used in the program, module, or procedure. It should be described in tabular format with columns for Name, Data Type, Description, and Attributes.
+Only include the Attributes column if necessary.
```
Add the following information about how to describe externally-described fields that are implicitly defined by the compiler:

If it is a field that was defined by the compiler from an externally described file, mention that in the description. For example if you have this file:

```
     A          R ORDERFMT
     A            ORDERID        6A
     A            QUANTITY       5P 0
     A            PRICE          9P 2
     A          K ORDERID
```

and this rpgle code
```rpgle
       dcl-f orders;
       dcl-s total int(10);

       begsr add_to_total;
          total += quantity * price;
       endsr;
```

Put this for the description when describing the global variables used by the subroutine:

| Name           | Data Type               | Description                                 | 
|----------------|------------------------ |---------------------------------------------|
| Price | packed(9:2), (5 bytes) | Externally-described field from file 'ORDERS', likely set by a READ operation from the file |

Commit:
Changes as per new format
